File: JANUARY/11-01-2026/85.rb
def maximal_rectangle(matrix)
    return 0 if matrix.empty?

    cols = matrix[0].length
    heights = Array.new(cols, 0)
    ans = 0

    matrix.each do |row|
        (0...cols).each do |j|
            heights[j] = row[j] == '1' ? heights[j] + 1 : 0
        end

        stack = []
        (0..cols).each do |i|
            h = (i == cols) ? 0 : heights[i]
            while !stack.empty? && h < heights[stack[-1]]
                height = heights[stack.pop]
                width = stack.empty? ? i : i - stack[-1] - 1
                ans = [ans, height * width].max
            end
            stack << i
        end
    end

    ans
end

File: JANUARY/11-01-2026/85.ts
function maximalRectangle(matrix: string[][]): number {
    if (matrix.length === 0) return 0;

    const cols = matrix[0].length;
    const heights: number[] = new Array(cols).fill(0);
    let ans = 0;

    for (const row of matrix) {
        for (let j = 0; j < cols; j++) {
            heights[j] = row[j] === '1' ? heights[j] + 1 : 0;
        }

        const stack: number[] = [];
        for (let i = 0; i <= cols; i++) {
            const h = i === cols ? 0 : heights[i];
            while (stack.length && h < heights[stack[stack.length - 1]]) {
                const height = heights[stack.pop()!];
                const width = stack.length ? i - stack[stack.length - 1] - 1 : i;
                ans = Math.max(ans, height * width);
            }
            stack.push(i);
        }
    }

    return ans;
}